### PR TITLE
Feature/cumulus 2407 data migration upserts

### PR DESCRIPTION
**Pull request by npauzenga**
**Summary:** Summary of changes

Addresses [CUMULUS-2407: Updated upsert and skip logic across PG data migrations for all data types](https://bugs.earthdata.nasa.gov/browse/CUMULUS-2407)

## Changes

* Replaced `UPDATE` queries in data-migration lambdas with `UPSERT` to allow newer records to overwrite old ones.
* Updates all data types in data-migration1 and data-migration2 to check the updated_at date of the dynamo record before upserting

## PR Checklist

- [x] Update CHANGELOG
- [x] Unit tests
- [ ] Adhoc testing
- [ ] Integration tests

### CODE REVIEW

File: lambdas/data-migration1/tests/test-rules.js
Comment (npauzenga) [inline lambdas/data-migration1/tests/test-rules.js:351]: I didn't want to change the Dynamo model for the sake of this test. The Rule Dynamo model implements its own version of `create` which sets the `updated_at` to `Date.now()`. That just means that I can't give it a different date when we create the Dynamo record.

File: lambdas/data-migration2/src/granulesAndFiles.ts
Comment (npauzenga) [inline lambdas/data-migration2/src/granulesAndFiles.ts:85]: I'm not 100% sure here but I think this is what we want to preserve the existing upsert logic where we're guarding against out-of-order writes.

This means that we will never update a record in the `running` state. That seems safest given the concerns but I want to flag it. 

An alternative might be to add `updated_at` to the `GranulesExecutions` JOIN table. We could check that date and if it's less than or equal to the dynamo record's updated_at, we could update:

`if (granulesExecutionsRecord.updated_at <= dynamoRecord.updated_at) { update the granule and granulesExecution }`

File: lambdas/data-migration2/src/pdrs.ts
Comment (npauzenga) [inline lambdas/data-migration2/src/pdrs.ts:62]: Same as above. PDRs also don't update if the state is `running`. The UPSERT logic is a little different than Granules though. The outcome is the same but there would need to be a different fix if we can't skip running records like this for some reason.

File: lambdas/data-migration1/src/async-operations.ts
Comment (markdboyd) [inline lambdas/data-migration1/src/async-operations.ts:48]: THANK YOU for updating the migration code to use the models 👍

File: lambdas/data-migration1/src/async-operations.ts
Comment (markdboyd) [inline lambdas/data-migration1/src/async-operations.ts:38]: We should probably replicate this logic across the other migrations so that they're using the `.get` method on the models, but we don't have to do it in this PR

File: lambdas/data-migration1/tests/test-async-operations.js
Comment (markdboyd) [inline lambdas/data-migration1/tests/test-async-operations.js:166]: Can we update this to use the model `get()`? I'd like to standardize on that instead of having the table name strings repeated throughout our code

File: lambdas/data-migration1/tests/test-collections.js
Comment (markdboyd) [inline lambdas/data-migration1/tests/test-collections.js:216]: Same: can we use collectionPgModel.get() here?

File: lambdas/data-migration1/tests/test-providers.js
Comment (markdboyd) [inline lambdas/data-migration1/tests/test-providers.js:317]:
```suggestion
test.serial('migrateProviderRecord updates an already migrated record if the updated timestamp on incoming record is newer', async (t) => {
```

File: lambdas/data-migration1/tests/test-providers.js
Comment (markdboyd) [inline lambdas/data-migration1/tests/test-providers.js:331]: Same: use the rule model `.get`

File: lambdas/data-migration1/tests/test-rules.js
Comment (markdboyd) [inline lambdas/data-migration1/tests/test-rules.js:25]: Definitely cleaner 👍

File: lambdas/data-migration1/tests/test-rules.js
Comment (markdboyd) [inline lambdas/data-migration1/tests/test-rules.js:283]:
```suggestion
test.serial('migrateRuleRecord updates an already migrated record if the updated timestamp on incoming record is newer', async (t) => {
```

File: lambdas/data-migration1/tests/test-rules.js
Comment (markdboyd) [inline lambdas/data-migration1/tests/test-rules.js:304]: Same: use rule PG model `.get`

File: lambdas/data-migration1/tests/test-rules.js
Comment (npauzenga) [reply lambdas/data-migration1/tests/test-rules.js:351]: 💥 I think @markdboyd handled this in https://github.com/nasa/cumulus/pull/2164/files?file-filters%5B%5D=.js

I will adjust here.

Comment (markdboyd) [reply lambdas/data-migration1/tests/test-rules.js:351]: Yeah I was just gonna say that. But what are you planning to change here? Would it be easier to wait until my PR's changes are merged?

File: lambdas/data-migration2/src/granulesAndFiles.ts
Comment (markdboyd) [reply lambdas/data-migration2/src/granulesAndFiles.ts:85]: Can you help me understand what we're trying to defend against here? For the parallel writes, we need the protection against out-of-order writes because the events can be received out of order. 

But the migrations should just take whatever is in Dynamo and copy it to PG, regardless of what status it is in or otherwise, unless I'm missing something.

There will be cases where at the time we migrate a granule, it is "running" in Dynamo, so copies to PG as "running". But then the execution finishes and the granule becomes "completed" in Dynamo and it **should** get parallel written to Dynamo and PG

File: lambdas/data-migration2/src/pdrs.ts
Comment (markdboyd) [reply lambdas/data-migration2/src/pdrs.ts:62]: Same question as above

File: lambdas/data-migration2/tests/test-executions.js
Comment (markdboyd) [inline lambdas/data-migration2/tests/test-executions.js:252]: Can we update this to use the execution PG model `.get`?

File: lambdas/data-migration2/tests/test-granulesAndFiles.js
Comment (markdboyd) [inline lambdas/data-migration2/tests/test-granulesAndFiles.js:422]: I don't know if we need to do a deepEqual comparison to the whole record. We only need to prove that the properties that should have been updated were changed to the correct value, so in this case the `updated_at` timestamp

File: lambdas/data-migration2/tests/test-pdrs.js
Comment (markdboyd) [inline lambdas/data-migration2/tests/test-pdrs.js:298]: I would explicitly set `updatedAt` on this first PDR to something that is older than the `Date.now()` timestamp used below. Obviously what you have here works but it's preferable to be a bit more explicit in the test setup

File: lambdas/data-migration2/tests/test-pdrs.js
Comment (markdboyd) [inline lambdas/data-migration2/tests/test-pdrs.js:311]: Remove the console logs

File: lambdas/data-migration2/tests/test-granulesAndFiles.js
Comment (markdboyd) [inline lambdas/data-migration2/tests/test-granulesAndFiles.js:404]: I would explicitly set `updatedAt` on this first granule to something that is older than the `Date.now()` timestamp on the `newerGranule`. Obviously what you have here works but it's preferable to be a bit more explicit in the test setup

File: lambdas/data-migration2/tests/test-pdrs.js
Comment (markdboyd) [inline lambdas/data-migration2/tests/test-pdrs.js:316]: Likewise, I don't think we need a full deepEqual comparison but just to prove that the updated_at timestamp is set correctly to prove that the record update happened

File: packages/db/tests/models/test-async-operation-model.js
Comment (markdboyd) [inline packages/db/tests/models/test-async-operation-model.js:62]: Change `updatedRule` to `updatedAsyncOperation`

File: packages/db/tests/models/test-async-operation-model.js
Comment (markdboyd) [inline packages/db/tests/models/test-async-operation-model.js:14]:
```suggestion
const testDbName = `async_operation_${cryptoRandomString({ length: 10 })}`;
```

File: CHANGELOG.md
Comment (markdboyd) [inline CHANGELOG.md:131]:
```suggestion
    - Changd data-migration1 and data-migration2 logic to only update already migrated records if the incoming record update has a newer timestamp
```

**Review by markdboyd — CHANGES_REQUESTED** [review]
Looks great! Mostly minor comments but a couple questions and requested changes

File: lambdas/data-migration1/tests/test-rules.js
Comment (npauzenga) [reply lambdas/data-migration1/tests/test-rules.js:351]: That might make sense. It's not a huge deal but I should be able to remove this line and set `updatedAt` when we create the Dynamo record in `await rulesModel.create(fakeRule);`. I'll double-check once your changes are merged though. The above PR will likely get in first anyway.

File: lambdas/data-migration2/src/granulesAndFiles.ts
Comment (npauzenga) [reply lambdas/data-migration2/src/granulesAndFiles.ts:85]: Yeah, so the issue here is the `upsert` logic in `GranulePgModel`. 

If I remove this line in the data-migration lambda that checks the status, we'll always try to upsert a granule if the record coming in is newer than the one already migrated. That's a problem because the `GranulePgModel.upsert` logic won't actually allow the record to be updated if it's in the `running` status. In that case we'll already have a `granulesExecutionsPgRecord` so we'll trigger the `whereNotExists` clause and prevent the upsert. 

Does that make sense? Happy to talk it through.

Comment (npauzenga) [reply lambdas/data-migration2/src/granulesAndFiles.ts:85]: Summing up our side conversation: we'll skip this `status` check in data-migration2 and handle the error that comes when an upsert fails due to a running granule.

Comment (npauzenga) [reply lambdas/data-migration2/src/granulesAndFiles.ts]: This commit:

https://github.com/nasa/cumulus/pull/2162/commits/6bead8bf8780e996500c39e0df452b90788f69f7

throws a new error type if the upsert didn't return any rows and we were therefor not able to update a granule during migration.

File: lambdas/data-migration2/src/pdrs.ts
Comment (npauzenga) [reply lambdas/data-migration2/src/pdrs.ts]: Same(ish) fix as above:

https://github.com/nasa/cumulus/pull/2162/commits/b57e0d8a7cb8e2b23ff7250774ab3fe3ee3e116a

We're not using the return of the upsert in `migratePdrRecord` the same way we do for Granules so this looks a little different. I still think it makes sense to throw the error here rather than further up the chain.

File: lambdas/data-migration1/tests/test-providers.js
Comment (npauzenga) [reply lambdas/data-migration1/tests/test-providers.js:317]: That's cleaner. Wording on date comparisons is tricky.

File: lambdas/data-migration2/tests/test-pdrs.js
Comment (npauzenga) [reply lambdas/data-migration2/tests/test-pdrs.js]: Ugh. Embarrassing.

https://github.com/nasa/cumulus/pull/2162/commits/cd3706e8988af761d806400b7603bef0dcb7fb37

File: packages/db/tests/models/test-async-operation-model.js
Comment (npauzenga) [reply packages/db/tests/models/test-async-operation-model.js]: Gross. This was (obviously) copy pasta.

https://github.com/nasa/cumulus/pull/2162/commits/ac551f1e053ed9dccc8b009a8fa0e0ab0aab177c

File: lambdas/data-migration1/src/async-operations.ts
Comment (npauzenga) [reply lambdas/data-migration1/src/async-operations.ts]: No problem, easy enough to do here:

https://github.com/nasa/cumulus/pull/2162/commits/5578e2474b3198a3b4640b1e81c74746198a6a42

File: lambdas/data-migration1/tests/test-async-operations.js
Comment (npauzenga) [reply lambdas/data-migration1/tests/test-async-operations.js]: https://github.com/nasa/cumulus/pull/2162/commits/7027bf1243c36dac3abc1052c1e7b28bdd60a305

File: lambdas/data-migration1/tests/test-collections.js
Comment (npauzenga) [reply lambdas/data-migration1/tests/test-collections.js]: https://github.com/nasa/cumulus/pull/2162/commits/7027bf1243c36dac3abc1052c1e7b28bdd60a305

File: lambdas/data-migration1/tests/test-providers.js
Comment (npauzenga) [reply lambdas/data-migration1/tests/test-providers.js]: https://github.com/nasa/cumulus/pull/2162/commits/7027bf1243c36dac3abc1052c1e7b28bdd60a305

File: lambdas/data-migration1/tests/test-rules.js
Comment (npauzenga) [reply lambdas/data-migration1/tests/test-rules.js]: https://github.com/nasa/cumulus/pull/2162/commits/7027bf1243c36dac3abc1052c1e7b28bdd60a305

File: lambdas/data-migration2/tests/test-executions.js
Comment (npauzenga) [reply lambdas/data-migration2/tests/test-executions.js]: https://github.com/nasa/cumulus/pull/2162/commits/7027bf1243c36dac3abc1052c1e7b28bdd60a305

File: lambdas/data-migration2/tests/test-granulesAndFiles.js
Comment (npauzenga) [reply lambdas/data-migration2/tests/test-granulesAndFiles.js]: https://github.com/nasa/cumulus/pull/2162/commits/171504648b8a39c26376eaa6d4931da6b181c2e2

File: lambdas/data-migration2/tests/test-pdrs.js
Comment (npauzenga) [reply lambdas/data-migration2/tests/test-pdrs.js]: https://github.com/nasa/cumulus/pull/2162/commits/171504648b8a39c26376eaa6d4931da6b181c2e2

File: lambdas/data-migration2/tests/test-pdrs.js
Comment (npauzenga) [reply lambdas/data-migration2/tests/test-pdrs.js:295]: Yup, makes sense.

https://github.com/nasa/cumulus/pull/2162/commits/5ee1061cf1b4832853e78a61cbbe3f0acfe0bbb7

File: lambdas/data-migration2/tests/test-granulesAndFiles.js
Comment (npauzenga) [reply lambdas/data-migration2/tests/test-granulesAndFiles.js:407]: 👍 https://github.com/nasa/cumulus/pull/2162/commits/5ee1061cf1b4832853e78a61cbbe3f0acfe0bbb7

File: lambdas/data-migration1/tests/test-async-operations.js
Comment (markdboyd) [inline lambdas/data-migration1/tests/test-async-operations.js:98]: THANK YOU 👍 🙌

File: lambdas/data-migration2/src/granulesAndFiles.ts
Comment (markdboyd) [inline lambdas/data-migration2/src/granulesAndFiles.ts:101]: So this error is thrown because the upsert from `upsertGranuleWithExecutionJoinRecord` affected no rows, but doesn't throw an error?

File: lambdas/data-migration2/tests/test-granulesAndFiles.js
Comment (markdboyd) [inline lambdas/data-migration2/tests/test-granulesAndFiles.js:394]: So we expect the upsert to affect no rows because `newerGranule` has a later `updatedAt` timestamp? I can't totally follow the test setup and how we guarantee the expected result

File: lambdas/data-migration2/src/granulesAndFiles.ts
Comment (npauzenga) [reply lambdas/data-migration2/src/granulesAndFiles.ts:102]: Right. If the `upsert` doesn't return rows, `upsertGranuleWithExecutionJoinRecord` will return an empty array. In that case it made sense to me to have `migrateGranuleRecord` throw an error rather than return undefined. It's caller expects either an error or a valid value (which we could obviously change but this seemed cleaner).

This reminded me to update the docstring:

https://github.com/nasa/cumulus/pull/2162/commits/2e26bd3f46bb0b706434a3d3c4e4eb7bc7b39fb7

File: lambdas/data-migration2/tests/test-granulesAndFiles.js
Comment (npauzenga) [reply lambdas/data-migration2/tests/test-granulesAndFiles.js:400]: Yeah, this test wasn't clear.

I've updated it explicitly set the `status` of the test granule. That's what will cause the upsert to return 0 rows and the migration to fail. Also added some commentary:

https://github.com/nasa/cumulus/pull/2162/commits/0f9cfdbcfd7fa541920454abeb4ff537917a79e5

File: lambdas/data-migration2/src/granulesAndFiles.ts
Comment (markdboyd) [reply lambdas/data-migration2/src/granulesAndFiles.ts:102]: I might make the content of this message (here and in PDRs) more explicit both for our own benefit and for users

```suggestion
    throw new PostgresUpdateFailed(`Upsert for granule ${record.granuleId} returned no rows. Record was not updated in the Postgres table.`);
```

Comment (npauzenga) [reply lambdas/data-migration2/src/granulesAndFiles.ts]: That's much clearer:

https://github.com/nasa/cumulus/pull/2162/commits/ea03d80a16a603b86afc5c4b484f2e56b6072881

File: lambdas/data-migration2/tests/test-granulesAndFiles.js
Comment (npauzenga) [reply lambdas/data-migration2/tests/test-granulesAndFiles.js:401]: Update: changed the wording slightly and added a similar explanation to the PDR migration test:

https://github.com/nasa/cumulus/pull/2162/commits/ea03d80a16a603b86afc5c4b484f2e56b6072881

**Review by markdboyd — APPROVED** [review]

File: lambdas/data-migration1/src/async-operations.ts
Comment (Jkovarik) [inline lambdas/data-migration1/src/async-operations.ts:41]: (I'll check this assertion later on, just commenting so it doesn't get dropped) -- we should carefully make sure the updatedAt values are identical on write.  I'm assuming that's part of this work.

File: lambdas/data-migration1/src/collections.ts
Comment (Jkovarik) [inline lambdas/data-migration1/src/collections.ts:38]: Nit: this comment seems clear from the code.

File: lambdas/data-migration1/src/providers.ts
Comment (Jkovarik) [inline lambdas/data-migration1/src/providers.ts:74]: See above

File: lambdas/data-migration1/src/providers.ts
Comment (Jkovarik) [inline lambdas/data-migration1/src/providers.ts:69]: Nit: could this logic be DRY'd up?    Not needed for this PR but something to consider.

**Review by Jkovarik — COMMENTED** [review]
Adding comments as I see other review approvals.  Nothing blocking, just wanted to dump comments.

File: lambdas/data-migration1/src/async-operations.ts
Comment (markdboyd) [reply lambdas/data-migration1/src/async-operations.ts:41]: It's not, but it's part of this PR: https://github.com/nasa/cumulus/pull/2164

File: lambdas/data-migration1/src/collections.ts
Comment (npauzenga) [reply lambdas/data-migration1/src/collections.ts]: Yeah, that's fair. It also doesn't explain WHY we're doing that so it's a pretty useless comment.

https://github.com/nasa/cumulus/pull/2162/commits/197e285aba1cb3a18dd5986dc5a2662455b6a453

File: lambdas/data-migration1/src/providers.ts
Comment (npauzenga) [reply lambdas/data-migration1/src/providers.ts]: https://github.com/nasa/cumulus/pull/2162/commits/197e285aba1cb3a18dd5986dc5a2662455b6a453

File: lambdas/data-migration1/src/async-operations.ts
Comment (Jkovarik) [reply lambdas/data-migration1/src/async-operations.ts:39]: Ah, nice 💯  - it's probably not that big a deal, but is this blocked merge on that then?

File: lambdas/data-migration1/src/providers.ts
Comment (npauzenga) [reply lambdas/data-migration1/src/providers.ts:69]: Yeah, there's definitely a lot of duplication across the data-migration lambdas and tests generally. The `migrateXRecord` functions and `migrateXs` basically all do the same thing and could probably be DRY'd up quite a lot.

It would be nice to clean that up. Because we don't really have an obvious (please correct me) module that migration helpers could live in, I'm a little hesitant to do it here due to the size of the refactor (mostly tests) though.

Would it be worth doing separately? I get stuck thinking that these won't need to be extended further but I've never once been right about that assumption 😄

File: lambdas/data-migration1/src/async-operations.ts
Comment (npauzenga) [reply lambdas/data-migration1/src/async-operations.ts:39]: It seems like merging wouldn't need to be blocked but Mark's PR would definitely need to be in before migrations are run. How does that feel?

File: lambdas/data-migration1/src/providers.ts
Comment (npauzenga) [reply lambdas/data-migration1/src/providers.ts:69]: Had a chat in Slack about this. There is definitely duplication here but _at this moment_ it seems like it would cost more time overall to address it as opposed to saving time by making these easier to extend/test. 

That doesn't mean we shouldn't refactor though. I'm just thinking we should address that if/when it comes up, mainly due to the sorta transient nature of these Lambdas.

Also, I get that you're referring to one block of code and I'm extrapolating to the whole Lambda 😄.  The catch with moving this one block to a shared module somewhere is that we don't really have a super obvious place for it so we'd likely need to create one. Totally doable but that pushes it out of the "just do it" mentality for me.

**Pull request merged**
